### PR TITLE
[Ruby] Correctly parse and use query parameters provided in nextlink format

### PR DESCRIPTION
**Pull request by vishrutshah**
Fixe for https://github.com/Azure/autorest/issues/1426

### CODE REVIEW

**Comment by vishrutshah** [comment]
@veronicagg & @sarangan12 Feel free to review this PR when you get a chance. Thanks!

File: src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb [hunk 1]
```diff
@@ -129,6 +129,21 @@
...
 
+  it 'should get multiple pages with fragmented next link' do
+    page = @client.paging.get_multiple_pages_fragment_next_link_as_lazy('1.6', 'test_user')
+    expect(page.is_a? PagingModule::Models::ProductResult)
+    expect(page.odatanext_link).not_to be_nil
+
+    all_item = page.get_all_items
+    expect(all_item.count).to eq(10)
+
+    # Get all items at once
```
Comment (veronicagg) [inline src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb]: >  # Get all items at once [](start = 2, length = 25)

feel free to split this into a separate test if you'd like, I've done that with others.

File: src/generator/AutoRest.Ruby.Azure/TemplateModels/AzureMethodTemplateModel.cs [hunk 2]
```diff
@@ -64,30 +64,39 @@ public bool IsPageable
         }
 
         /// <summary>
-        /// Returns invocation string for next method async.
+        /// Returns string for next method assigned to the Page.
```
Comment (veronicagg) [inline src/generator/AutoRest.Ruby.Azure/TemplateModels/AzureMethodTemplateModel.cs]: >string [](start = 20, length = 6)

should we say "Returns the code which assigns the page.next_method" or something like that?

File: src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb [hunk 3]
```diff
@@ -129,6 +129,21 @@
     expect(items.count).to eq(10)
   end
 
+  it 'should get multiple pages with fragmented next link' do
```
Comment (veronicagg) [inline src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb]: should we be adding some more cases in swagger (paging.json) to account for paramater grouping with paging, and maybe some of the other cases we tried, to exercise our code?

File: src/generator/AutoRest.Ruby.Azure/TemplateModels/AzureMethodTemplateModel.cs [hunk 4]
```diff
@@ -97,13 +106,14 @@ public string InvokeNextMethodAsync()
                         builder.AppendLine(string.Format(CultureInfo.InvariantCulture, "{0} = {1}", param.Name, argumentName));
                     }
                 }
+                builder.Outdent();
             }
 
-            IList<string> headerParams = nextMethod.Parameters.Where(p => (p.Location == ParameterLocation.Header || p.Location == ParameterLocation.None) && !p.IsConstant && p.ClientProperty == null).Select(p => p.Name).ToList();
-            headerParams.Add("custom_headers");
```
Comment (veronicagg) [inline src/generator/AutoRest.Ruby.Azure/TemplateModels/AzureMethodTemplateModel.cs]: are we losing "custom_headers"?

**Comment by veronicagg** [comment]
:clock1:

File: src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb (hunk 1 above)
Comment (vishrutshah) [inline src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb]: Sure.

File: src/generator/AutoRest.Ruby.Azure/TemplateModels/AzureMethodTemplateModel.cs (hunk 2 above)
Comment (vishrutshah) [inline src/generator/AutoRest.Ruby.Azure/TemplateModels/AzureMethodTemplateModel.cs]: How about 
```
Returns Ruby code as string which sets `next_method` property of the page with the respective next paging method and performs parameter re-assignment when required (ex. parameter grouping cases)
```

File: src/generator/AutoRest.Ruby.Azure/TemplateModels/AzureMethodTemplateModel.cs (hunk 4 above)
Comment (vishrutshah) [inline src/generator/AutoRest.Ruby.Azure/TemplateModels/AzureMethodTemplateModel.cs]: Nope but we don't need that explicit addition as `nextMethodTemplateModel.MethodParameterInvocation` includes it as shown in below example

```
    def get_multiple_pages_fragment_next_link_as_lazy(api_version, tenant, custom_headers = nil)
      response = get_multiple_pages_fragment_next_link_async(api_version, tenant, custom_headers).value!
      unless response.nil?
        page = response.body
        page.next_method = Proc.new do |next_link|
          next_fragment_async(api_version, tenant, next_link, custom_headers)
        end
        page
      end
    end
```

File: src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb (hunk 3 above)
Comment (vishrutshah) [inline src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb]: Added a test in AutoRest to cover paging method with operation name and grouped parameters. 

@tbombach Could you also please review the changes in second commit to make sure i did things right? 

Thanks

File: src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb [hunk 5]
```diff
@@ -129,6 +129,30 @@
     expect(items.count).to eq(10)
   end
 
+  it 'should get multiple pages with fragmented next link - lazy' do
+    page = @client.paging.get_multiple_pages_fragment_next_link_as_lazy('1.6', 'test_user')
```
Comment (sarangan12) [inline src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb]: I see 1.6 & test_user are used in atleast 2 cases and most probably in more in the future. Can we define them in one place and use it everywhere

Comment (vishrutshah) [inline src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb]: Sure that sounds perfect. will update it in next iteration.

Comment (vishrutshah) [inline src/generator/AutoRest.Ruby.Azure.Tests/RspecTests/paging_spec.rb]: updated.

**Comment by vishrutshah** [comment]
Thanks for the review!!

**Pull request merged**